pamelalozano/src/main/java/com/google/sps/servlets/NicknameServlet.java

```diff
@@ -0,0 +1,93 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/nickname")
+public class NicknameServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("text/html");
+    PrintWriter out = response.getWriter();
+    out.println("<h1>Set Nickname</h1>");
+
+    UserService userService = UserServiceFactory.getUserService();
+    if (userService.isUserLoggedIn()) {
+      String nickname = getUserNickname(userService.getCurrentUser().getUserId());
+
+      out.println("<p>Set your nickname here:</p>");
+      out.println("<form method=\"POST\" action=\"/nickname\">");
+      out.println("<input name=\"nickname\" value=\"" + nickname + "\" />");
+      out.println("<br/>");
+      out.println("<button>Submit</button>");
+      out.println("</form>");
+    } else {
+      String loginUrl = userService.createLoginURL("/nickname");
+      out.println("<p>Login <a href=\"" + loginUrl + "\">here</a>.</p>");
+    }
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/nickname");
+      return;
+    }
+
+    String nickname = request.getParameter("nickname");
+    String id = userService.getCurrentUser().getUserId();
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity entity = new Entity("UserInfo", id);
+    entity.setProperty("id", id);
+    entity.setProperty("nickname", nickname);
+    // The put() function automatically inserts new data or updates existing data based on ID
+    datastore.put(entity);
+
+    response.sendRedirect("/");
+  }
+
+  /**
+   * Returns the nickname of the user with id, or empty String if the user has not set a nickname.
+   */
+  private String getUserNickname(String id) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query =
```

Comment:
I think you can create a UserInfo Key directly like `Key userInfoKey = KeyFactory.createKey("UserInfo", id)` and do `Entity entity = datastore.get(key)`. 

I don't think it really makes much of a difference in this case since "id" is indexed. Let me know if this works if you try it out!

Reply:
Okay it took me a little long to understand what you meant and it kept throwing an error but I did a little research and the function know looks like this and I tested it it works fine:
private String getUserNickname(String id) {
    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
    Entity entity = null;
   
try {
        Key userInfoKey = KeyFactory.createKey("UserInfo", id);
        entity = datastore.get(userInfoKey);
    } catch(Exception e){
        System.out.println(e);
    }
   
 if (entity == null) {
      return null;
    }
    
   String nickname = (String) entity.getProperty("nickname");
    return nickname;
  }

Reply:
Nice!